### PR TITLE
Added buttons to Unity

### PR DESCRIPTION
**Pull request by fenwikk**

### CODE REVIEW

File: Unity Example/Assets/Discord RPC/Scripts/DiscordManager.cs [hunk 1]
```diff
@@ -182,7 +182,6 @@ public void Initialize()
 
         //Assign the instance
         _instance = this;
-        DontDestroyOnLoad(this);
```
Comment (Lachee) [inline Unity Example/Assets/Discord RPC/Scripts/DiscordManager.cs:185]: Why was this removed here?

File: Unity Example/Assets/Discord RPC/Scripts/DiscordManager.cs [hunk 2]
```diff
@@ -152,18 +153,17 @@ private static void CreateNewManager()
...
+	/// <summary>
+	/// Initializes the discord client if able to. Wont initialize if <see cref="active"/> is false, we are not in playmode, we already have a instance or we already have a client.
+	/// <para>This function is empty unless UNITY_WSA || UNITY_WSA_10_0 || UNITY_STANDALONE) && !DISABLE_DISCORD is meet.</para>
+	/// </summary>
+	public void Initialize()
+	{
 #if (UNITY_WSA || UNITY_WSA_10_0 || UNITY_STANDALONE) && !DISABLE_DISCORD
 
         if (!active) return;                //Are we allowed to be active?
-        if (!Application.isPlaying) return; //We are not allowed to initialize while in the editor.
```
Comment (Lachee) [inline Unity Example/Assets/Discord RPC/Scripts/DiscordManager.cs:166]: We don't want this to be running in the editor generally speaking. Unity does not behave well with native pipes.
Can you confirm that it doesn't keep the Presence once its out of play mode?

File: Unity Example/Assets/Discord RPC/Example/Rich Presence/Example_RichPresence.cs [hunk 3]
```diff
@@ -33,7 +33,7 @@ private void Start()
 			DiscordManager.current.events.OnPresenceUpdate.AddListener((message) =>
 			{
 				Debug.Log("Received a new presence! Current App: " + message.ApplicationID + ", " + message.Name);
-				UpdateFields(new DiscordPresence(message.Presence));
+				UpdateFields(new DiscordPresence(message.Presence.ToRichPresence()));
```
Comment (Lachee) [inline Unity Example/Assets/Discord RPC/Example/Rich Presence/Example_RichPresence.cs:36]: This should be implicitly cast. A presence is a presence. 
However the update fields need to account for the fact that discord doesn't return button information correctly at the same time in the OnPresenceUpdate event.

File: DiscordRPC/RichPresence.cs [hunk 4]
```diff
@@ -920,6 +955,41 @@ internal override bool Matches(RichPresence other)
         {
             return presesnce != null;
         }
+
+        /// <summary>
+        /// Converts this RichPresence to BaseRichPresence
+        /// </summary>
+        /// <returns></returns>
+        public BaseRichPresence ToBaseRichPresence()
```
Comment (Lachee) [inline DiscordRPC/RichPresence.cs:963]: This would be almost never used. 
BasePresence is basically just for returned data back from discord.

File: DiscordRPC/DiscordRPC.csproj [hunk 5]
```diff
@@ -6,7 +6,7 @@
     <DocumentationFile>bin\$(Configuration)\$(TargetFramework)\$(AssemblyName).xml</DocumentationFile>
   </PropertyGroup>
   <ItemGroup>
-    <PackageReference Include="Newtonsoft.Json" Version="12.0.2" />
+    <PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
```
Comment (Lachee) [inline DiscordRPC/DiscordRPC.csproj:9]: Is a Newtonsoft.Json update required? This is a major update too, and for compatibility reasons I try to avoid updating this library.

**Review by Lachee — COMMENTED** [review]
This looks good, just a few comments I need cleared up and some discussion first.
Thank you for your interest in the library and contributions :)

File: DiscordRPC/DiscordRPC.csproj (hunk 5 above)
Comment (fenwikk) [reply DiscordRPC/DiscordRPC.csproj]: yeah fixed that

File: DiscordRPC/RichPresence.cs (hunk 4 above)
Comment (fenwikk) [reply DiscordRPC/RichPresence.cs]: gotcha removed it

File: Unity Example/Assets/Discord RPC/Scripts/DiscordManager.cs (hunk 2 above)
Comment (fenwikk) [reply Unity Example/Assets/Discord RPC/Scripts/DiscordManager.cs]: yeah fixed that

File: Unity Example/Assets/Discord RPC/Scripts/DiscordManager.cs (hunk 1 above)
Comment (fenwikk) [reply Unity Example/Assets/Discord RPC/Scripts/DiscordManager.cs]: yeah fixed that

File: Unity Example/Assets/Discord RPC/Example/Rich Presence/Example_RichPresence.cs (hunk 3 above)
Comment (fenwikk) [reply Unity Example/Assets/Discord RPC/Example/Rich Presence/Example_RichPresence.cs:36]: dont really know how to do that just yet

**Comment by Lachee** [comment]
ok ill look into this before Monday and hopefully merge this for you. Been super busy :)

File: Unity Example/Assets/Discord RPC/Scripts/Presence/DiscordPresence.cs [hunk 6]
```diff
@@ -151,6 +184,15 @@ public DiscordRPC.RichPresence ToRichPresence()
...
 		}
 
+		if ((firstButton != null && !firstButton.IsEmpty()) || (secondButton != null && !secondButton.IsEmpty()))
+		{
+			presence.Buttons = new DiscordRPC.Button[]
+			{
+				new DiscordRPC.Button { Label = firstButton.label, Url = firstButton.url},
+				new DiscordRPC.Button { Label = secondButton.label, Url = secondButton.url}
+			};
+		}
```
Comment (Lachee) [inline Unity Example/Assets/Discord RPC/Scripts/Presence/DiscordPresence.cs:194]: This will have to change for this PR to be accepted.
The buttons are an array, and thus should be listed in the inspector as an array. 

If discord only accepts 2 buttons, then this should reflect the base library and not be custom implemented in the unity package.

**Review by Lachee — CHANGES_REQUESTED** [review]

**Comment by fenwikk** [comment]
Is there anything that needs to be added?

File: Unity Example/Assets/Discord RPC/Scripts/Presence/DiscordPresence.cs [hunk 7]
```diff
@@ -98,11 +106,33 @@ public DiscordPresence(DiscordRPC.BaseRichPresence presence)
...
+					{
+						label = presence.Buttons[0].Label,
+						url = presence.Buttons[0].Url,
+					},
+					new DiscordButton()
+					{
+						label = presence.Buttons[1].Label,
+						url = presence.Buttons[1].Url,
+					}
+				};
```
Comment (Lachee) [inline Unity Example/Assets/Discord RPC/Scripts/Presence/DiscordPresence.cs:123]: This shouldn't be a new array of only 2 items. Copy the entire presence.Buttons object.

Buttons is an array with an **unbounded length**. If this is incorrect, please create a PR to fix this in the base library. The Unity package should match functionality of the base library.

File: Unity Example/Assets/Discord RPC/Scripts/Presence/DiscordPresence.cs [hunk 8]
```diff
@@ -151,6 +182,15 @@ public DiscordRPC.RichPresence ToRichPresence()
 			if (endTime.IsValid()) presence.Timestamps.End = endTime.GetDateTime();
 		}
 
+		if (buttons.Length > 0)
+		{
+			presence.Buttons = new DiscordRPC.Button[]
+			{
+				new DiscordRPC.Button { Label = buttons[0].label, Url = buttons[0].url},
+				new DiscordRPC.Button { Label = buttons[1].label, Url = buttons[1].url}
+			};
```
Comment (Lachee) [inline Unity Example/Assets/Discord RPC/Scripts/Presence/DiscordPresence.cs:191]: This shouldn't be a new array of fixed length. Buttons can be any length. See other review.

**Review by Lachee — CHANGES_REQUESTED** [review]

File: Unity Example/Assets/Discord RPC/Scripts/Presence/DiscordPresence.cs (hunk 7 above)
Comment (fenwikk) [reply Unity Example/Assets/Discord RPC/Scripts/Presence/DiscordPresence.cs]: fixed

File: Unity Example/Assets/Discord RPC/Scripts/Presence/DiscordPresence.cs (hunk 8 above)
Comment (fenwikk) [reply Unity Example/Assets/Discord RPC/Scripts/Presence/DiscordPresence.cs]: fixed

File: Unity Example/Assets/Discord RPC/Scripts/Presence/DiscordPresence.cs (hunk 6 above)
Comment (fenwikk) [reply Unity Example/Assets/Discord RPC/Scripts/Presence/DiscordPresence.cs]: fixed

**Review by Lachee — APPROVED** [review]
